update-client-hub/modules/pal-psa/source/arm_uc_pal_psa_implementation.c

```diff
@@ -0,0 +1,548 @@
+                                              &firmware_update_ctx_nonsecure.psa_fwu.dependency_uuid,
+                                              &firmware_update_ctx_nonsecure.psa_fwu.dependency_version);
+        if (status != PSA_SUCCESS && status != PSA_SUCCESS_REBOOT) {
+            UC_PAAL_ERR_MSG("psa_fwu_install() failed: %d\r\n", status);
+            result.code = FIRM_ERR_ACTIVATE;
+            return result;
+        }
+
+        /* signal done */
+        arm_uc_pal_psa_signal_internal(ARM_UC_PAAL_EVENT_ACTIVATE_DONE);
+    
+        result.code = ERR_NONE;
+    }
+
+    return result;
+}
+
+/**
+ * @brief Get firmware details for the firmware image in the slot passed.
+ * @details This call populates the passed details struct with information
+ *          about the firmware image in the slot passed. Only the fields
+ *          marked as supported in the capabilities bitmap will have valid
+ *          values.
+ *
+ * @param details Pointer to firmware details struct to be populated.
+ * @return Returns ERR_NONE on accept, and signals the event handler with
+ *         either DONE or ERROR when complete.
+ *         Returns ERR_INVALID_PARAMETER on reject, and no signal is sent.
+ */
+arm_uc_error_t ARM_UC_PAL_PSA_Mcuboot_GetFirmwareDetails(
+    uint32_t slot_id,
+    arm_uc_firmware_details_t* details)
+{
+    UC_PAAL_TRACE("ARM_UC_PAL_PSA_Mcuboot_GetFirmwareDetails");
+
+    (void) slot_id;
+    (void) details;
+
+    /* This function is not used for MCUBOOT. */
+
+    arm_uc_error_t result = { .code = ERR_NOT_READY };
+
+    return result;
+}
+
+/*****************************************************************************/
+
+arm_uc_error_t ARM_UC_PAL_PSA_Mcuboot_GetActiveDetails(arm_uc_firmware_details_t* details)
+{
+    UC_PAAL_TRACE("ARM_UC_PAL_PSA_Mcuboot_GetActiveDetails");
+
+    arm_uc_error_t result = { .code = ERR_INVALID_PARAMETER };
+
+    if (details) {
+
+        /* Clear for clean */
+        memset(details, 0x00, sizeof(arm_uc_firmware_details_t));
+
+        /* parse MCUBOOT header and get hash from TLV struct and total size */
+        arm_uc_hash_t header_hash = { 0 };
+        uint32_t total_size = 0;
+
+        result = 
+            arm_uc_pal_psa_get_hash_from_header(arm_uc_flashiap_active_read,
+                                                           MBED_CONF_UPDATE_CLIENT_APPLICATION_DETAILS,
+                                                           &header_hash,
+                                                           &total_size);
+        if (result.code != ERR_NONE) {
+            return result;
+        }
+
+        /* use MCUBOOT hash to lookup firmware details in KCM */
+        result = arm_uc_pal_psa_get_kcm_details(&header_hash, details);
+
+        /* if no details were found in KCM */
+        if (result.error != ERR_NONE) {
+
+            /* calculate hash from active image */
+            result = arm_uc_pal_psa_calculate_hash(arm_uc_flashiap_active_read,
```

Comment:
I added this to the MCUBoot version to handle the corner case where one wanted to do a delta update as the very first update. In that scenario, the KCM would be empty and the update client would not be able to verify that the delta image matches the image it is supposed to be applied to. 

For the PSA implementation, I would expect the secure side, e.g., TFM, to do the delta update since reading from memory is not guaranteed with PSA.

Reply:
TF-M's `psa_fwu_write()` will do whole candidate area erase at the first call. This causes delta update cannot support. Maybe add some guard from enabling delta update for PSA.
https://github.com/OpenNuvoton/trusted-firmware-m/blob/149271c63a448b0d690b41ae62dfbf3f36abcf72/secure_fw/partitions/firmware_update/bootloader/mcuboot/tfm_mcuboot_fwu.c#L187-L189

Reply:
Not clear about delta update. It seems Update Client needs to read active firmware plus delta update patch to generate intended new firmware for update. For non-secure firmware update, it is possible because non-secure active firmware can read. But for secure or combined firmware update, read is prohibited as pointed out.